### PR TITLE
bcrypt

### PR DESCRIPTION
**Pull request by Avetiq30**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -29,6 +29,7 @@
     "@nestjs/platform-express": "^10.0.0",
     "@nestjs/typeorm": "^10.0.0",
     "@typegoose/typegoose": "^11.5.0",
+    "@types/bcrypt": "^5.0.0",
```
Comment (VahagShahinyan) [inline package.json:32]: Es teghapoxi ```devDependencies```

File: src/auth/auth.service.ts [hunk 2]
```diff
@@ -11,10 +12,13 @@ export class AuthService {
...
-      return { token };
+    if (user) {
+      const isPasswordValid =  bcrypt.compare(password, user.password);
+      if (isPasswordValid) {
+        const payload = { email: user.email, sub: user.password };
+        const token = this.jwtService.sign(payload);
+        return { token };
+      }
     }
     return null;
```
Comment (VahagShahinyan) [inline src/auth/auth.service.ts:23]: change 
```js 
 if (!user) {
        return;
    }
    const isPasswordValid = bcrypt.compare(password, user.password);
    if (!isPasswordValid) {
        return;
    }
    const payload = { email: user.email, sub: user.password };
    const token = this.jwtService.sign(payload);
    return { token };
    ```

File: src/auth/auth.controller.ts [hunk 3]
```diff
@@ -21,6 +21,6 @@ export class AuthController {
     if (!result) {
       throw new HttpException('Forbidden', HttpStatus.FORBIDDEN);
```
Comment (VahagShahinyan) [inline src/auth/auth.controller.ts:22]: Update Error message

File: src/auth/auth.service.ts [hunk 4]
```diff
@@ -11,10 +12,13 @@ export class AuthService {
 
   async validateUser(email: string, password: string): Promise<any> {
```
Comment (VahagShahinyan) [inline src/auth/auth.service.ts:13]: Why validateUser should return token ?

Comment (VahagShahinyan) [inline src/auth/auth.service.ts:13]: Refactor validateUser method

**Comment by VahagShahinyan** [comment]
- Check API functionality 
- Run Prettier and Eslint
- Add Husky
- Add integration test

File: src/products/products.service.ts [hunk 5]
```diff
@@ -22,23 +23,19 @@ export class ProductsService {
...
+  async getProductById(id: string): Promise<ProductModel | null> {
+    return this.productModel.findById(id).exec();
   }
 
   async updateProduct(
     id: string,
     updateProductDto: UpdateProductDto,
-  ): Promise<any> {
+  ): Promise<ProductModel> {
     const updatedProduct = await this.productModel
```
Comment (VahagShahinyan) [inline src/products/products.service.ts:39]: Delete await

File: test/auth.controller.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    let app :INestApplication
+    let userService: UserService
+
+    beforeEach(async()=>{
+        await userService.deleteAll()
+    })
+    afterAll(async()=>{
+        await userService.deleteAll()
+    })
+
```
Comment (VahagShahinyan) [inline test/auth.controller.e2e-spec.ts:19]: duplicate function

File: test/auth.controller.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        app =  moduleFixture.createNestApplication()
+        userService = moduleFixture.get<UserService>(UserService)
+        await app.init()
+    })
+    afterAll(async ()=>{
+        await app.close()
+    })
+
+    describe('When trying login',()=>{
```
Comment (VahagShahinyan) [inline test/auth.controller.e2e-spec.ts:33]: change  `When trying login` to `When trying login in `

File: test/auth.controller.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+         const loginData = {
+             email: "test@mail.ru",
+             password: "test123"
+         }
+        const user: UserModel  =  {
+            name:'test',
+            lastname:"lasttest",
+            email:loginData.email,
+            password:loginData.password,
+        }
```
Comment (VahagShahinyan) [inline test/auth.controller.e2e-spec.ts:45]: Move to separate file, for example create userHelper.ts file

File: test/product.controller.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
+import { Test, TestingModule} from "@nestjs/testing"
+import { HttpStatus, INestApplication} from "@nestjs/common"
+import * as request from "supertest"
+import { AppModule } from "../src/app.module"
+import {ProductsService} from '../src/products/products.service'
+import { CreateProductDto, UpdateProductDto } from "src/products/dto/product.dto"
+
```
Comment (VahagShahinyan) [inline test/product.controller.e2e-spec.ts:7]: delete  line

File: test/product.controller.e2e-spec.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    beforeAll(async () => {
+        const moduleFixture: TestingModule = await Test.createTestingModule({
+          imports: [AppModule],
+        }).compile()
+    
+        app = moduleFixture.createNestApplication();
+        productService = moduleFixture.get<ProductsService>(ProductsService);
+        await app.init();
+      })
+      afterAll(async()=>{
```
Comment (VahagShahinyan) [inline test/product.controller.e2e-spec.ts:34]: duplicate function

File: test/product.controller.e2e-spec.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        await app.close()
+      })
+ 
+      describe('Created product', ()=>{
```
Comment (VahagShahinyan) [inline test/product.controller.e2e-spec.ts:38]: Change `Created product` to `When trying to create product `

File: test/product.controller.e2e-spec.ts [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        it('should be access',async()=>{
+          
+            const  createProduct:CreateProductDto = {
```
Comment (VahagShahinyan) [inline test/product.controller.e2e-spec.ts:41]: Move to separate file, for example create productHelper.ts file

File: test/product.controller.e2e-spec.ts [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+            expect(response.body.name).toBe(createProduct.name)
+            expect(response.body.description).toBe(createProduct.description)
+            expect(response.body.price).toBe(createProduct.price)
+            expect(response.body.category._id).toBe(createProduct.categoryId)
+
+           
+        })
+      })
+
+      describe('Update a product', ()=>{
```
Comment (VahagShahinyan) [inline test/product.controller.e2e-spec.ts:61]: Change to `When trying update a product`

File: test/product.controller.e2e-spec.ts [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+            }
+            
+            const response = await request(app.getHttpServer())
+            .put(`/products/${createdProduct.id}`)
+            .send(updateProductDto)
+            .expect(HttpStatus.OK)
+
+            expect(response.body).toMatchObject(updateProductDto)
+
+            const updatedProduct = await productService.getProductById(createdProduct.id)
```
Comment (VahagShahinyan) [inline test/product.controller.e2e-spec.ts:85]: delete

File: test/product.controller.e2e-spec.ts [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+                name:'updateTest',
+                price:299
+            }
+            
+            const response = await request(app.getHttpServer())
+            .put(`/products/${createdProduct.id}`)
+            .send(updateProductDto)
+            .expect(HttpStatus.OK)
+
+            expect(response.body).toMatchObject(updateProductDto)
```
Comment (VahagShahinyan) [inline test/product.controller.e2e-spec.ts:83]:
```js 
expect(response.body._id).toBe(createdProduct._id)
expect(response.body.name).toBe(updateProductDto.name)
expect(response.body.price).toBe(updateProductDto.price)
expect(response.body.categoryId).toBe(createdProduct.categoryId)
expect(response.body.description).toBe(createdProduct.description)

File: test/product.controller.e2e-spec.ts [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+            const updatedProduct = await productService.getProductById(createdProduct.id)
+           
+            
+            expect(updatedProduct.name).toBe(updateProductDto.name)
```
Comment (VahagShahinyan) [inline test/product.controller.e2e-spec.ts:88]: delete

File: test/user.controller.e2e-spec.ts [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+    beforeAll(async ()=>{
+        const moduleFixture: TestingModule = await Test.createTestingModule({
+            imports:[AppModule],
+        }).compile()
+        
+        app = moduleFixture.createNestApplication();
+        userService = moduleFixture.get<UserService>(UserService);
+        await app.init()
+    })
+    afterAll(async () => {
```
Comment (VahagShahinyan) [inline test/user.controller.e2e-spec.ts:27]: duplicate function

File: test/user.controller.e2e-spec.ts [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+        await app.close();
+      })
+    
+describe('When registering a user',()=>{
```
Comment (VahagShahinyan) [inline test/user.controller.e2e-spec.ts:31]: change `When trying registering a new user`

File: test/user.controller.e2e-spec.ts [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+    it('should register a user success',async ()=>{
```
Comment (VahagShahinyan) [inline test/user.controller.e2e-spec.ts:32]: change `should be success`

File: test/user.controller.e2e-spec.ts [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+            email: "testing@mail.ru",
+            password: "test123"
+        }
+       const userData: UserModel  =  {
+           name:'test',
+           lastname:"lasttest",
+           email:loginData.email,
+           password:loginData.password,
+       }
+       const hashedPassword = await bcrypt.hash(userData.password, 10);
```
Comment (VahagShahinyan) [inline test/user.controller.e2e-spec.ts:44]: move to service method and use it

File: src/auth/auth.service.ts [hunk 21]
```diff
@@ -9,13 +10,23 @@ export class AuthService {
...
     const user = await this.userService.findUserByEmail(email);
-    if (user && user.password === password) {
-      const payload = { email: user.email, sub: user.password };
-      const token = this.jwtService.sign(payload);
-      return { token };
+    if (!user) {
+      return false;
     }
-    return null;
+    return bcrypt.compare(password, user.password);
```
Comment (VahagShahinyan) [inline src/auth/auth.service.ts:18]: Add ``comparePassword()``  method in UserService and use it in this line

File: test/user.controller.e2e-spec.ts [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+         ...userData,
+         password: hashedPassword, 
+       };
+       
+       const response = await request(app.getHttpServer())
+       .post('/user/register')
+       .send(userData)
+       expect(HttpStatus.OK)
+       
+       const isPasswordMatch = await bcrypt.compare(userData.password, response.body.password);
```
Comment (VahagShahinyan) [inline test/user.controller.e2e-spec.ts:55]: Use `UserService.comparePassword()` method

File: test/user.controller.e2e-spec.ts [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+      expect(isPasswordMatch).toBe(true)
+      expect(response.body.name).toBe(user.name)
+      expect(response.body.lastname).toBe(user.lastname)
+      expect(response.body.email).toBe(user.email)
+
```
Comment (VahagShahinyan) [inline test/user.controller.e2e-spec.ts:60]: ``expect(response.body.password).toBe(user.password)``

File: test/user.controller.e2e-spec.ts [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+       const response = await request(app.getHttpServer())
+       .post('/user/register')
+       .send(userData)
+       expect(HttpStatus.OK)
+       
+       const isPasswordMatch = await bcrypt.compare(userData.password, response.body.password);
+      expect(isPasswordMatch).toBe(true)
+      expect(response.body.name).toBe(user.name)
+      expect(response.body.lastname).toBe(user.lastname)
+      expect(response.body.email).toBe(user.email)
```
Comment (VahagShahinyan) [inline test/user.controller.e2e-spec.ts:59]: delete

File: test/user.controller.e2e-spec.ts [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+         password: hashedPassword, 
+       };
+       
+       const response = await request(app.getHttpServer())
+       .post('/user/register')
+       .send(userData)
+       expect(HttpStatus.OK)
+       
+       const isPasswordMatch = await bcrypt.compare(userData.password, response.body.password);
+      expect(isPasswordMatch).toBe(true)
```
Comment (VahagShahinyan) [inline test/user.controller.e2e-spec.ts:56]: delete

**Review by VahagShahinyan — COMMENTED** [review]
- Run `eslint` and `prettier` before open PR
- Create `product-create.dto.ts ` and `product-update.dto.ts`
- Please follow REST API standard to create product api
- Include a test case for every condition present in the service or controller.

File: test/product.controller.e2e-spec.ts (hunk 9 above)
Comment (VahagShahinyan) [reply test/product.controller.e2e-spec.ts:7]: change ```from "src/products/dto/product.dto"``` to ```from "../src/products/dto/product.dto"```

File: test/user.controller.e2e-spec.ts (hunk 24 above)
Comment (Avetiq30) [reply test/user.controller.e2e-spec.ts:59]: ++

File: test/user.controller.e2e-spec.ts (hunk 23 above)
Comment (Avetiq30) [reply test/user.controller.e2e-spec.ts:60]: ++

File: test/user.controller.e2e-spec.ts (hunk 25 above)
Comment (Avetiq30) [reply test/user.controller.e2e-spec.ts:56]: ++

File: test/user.controller.e2e-spec.ts (hunk 19 above)
Comment (Avetiq30) [reply test/user.controller.e2e-spec.ts:32]: ++

File: test/user.controller.e2e-spec.ts (hunk 17 above)
Comment (Avetiq30) [reply test/user.controller.e2e-spec.ts:27]: ++

File: test/product.controller.e2e-spec.ts (hunk 16 above)
Comment (Avetiq30) [reply test/product.controller.e2e-spec.ts:88]: ++

File: test/product.controller.e2e-spec.ts (hunk 14 above)
Comment (Avetiq30) [reply test/product.controller.e2e-spec.ts:85]: ++

File: test/product.controller.e2e-spec.ts (hunk 15 above)
Comment (Avetiq30) [reply test/product.controller.e2e-spec.ts:83]: ++

File: test/product.controller.e2e-spec.ts (hunk 13 above)
Comment (Avetiq30) [reply test/product.controller.e2e-spec.ts:61]: ++

File: test/product.controller.e2e-spec.ts (hunk 11 above)
Comment (Avetiq30) [reply test/product.controller.e2e-spec.ts:38]: ++

File: test/product.controller.e2e-spec.ts (hunk 10 above)
Comment (Avetiq30) [reply test/product.controller.e2e-spec.ts:34]: ++

File: test/product.controller.e2e-spec.ts (hunk 9 above)
Comment (Avetiq30) [reply test/product.controller.e2e-spec.ts:7]: ++

File: test/auth.controller.e2e-spec.ts (hunk 6 above)
Comment (Avetiq30) [reply test/auth.controller.e2e-spec.ts:19]: ++

File: test/auth.controller.e2e-spec.ts (hunk 7 above)
Comment (Avetiq30) [reply test/auth.controller.e2e-spec.ts:33]: ++

**Pull request merged**